File: backend/src/services/llmService.js
import { HfInference } from '@huggingface/inference';
import dotenv from 'dotenv';

dotenv.config();

class LlmService {
  constructor() {
    this.hf = new HfInference(process.env.HUGGINGFACE_API_KEY);

    // Sử dụng Qwen2.5-7B-Instruct - Model free tốt nhất cho tiếng Việt
    this.model = 'Qwen/Qwen2.5-7B-Instruct';

    this.systemPrompt = `Bạn là trợ lý bán hàng laptop thông minh của một cửa hàng thương mại điện tử.

NHIỆM VỤ:
1. Tư vấn và giới thiệu sản phẩm laptop dựa trên thông tin được cung cấp
2. Trả lời câu hỏi về thông số kỹ thuật, giá cả, so sánh sản phẩm
3. Gợi ý sản phẩm phù hợp với nhu cầu và ngân sách khách hàng
4. Giải thích các thuật ngữ kỹ thuật một cách dễ hiểu

QUY TẮC QUAN TRỌNG:
- LUÔN dựa vào THÔNG TIN SẢN PHẨM được cung cấp trong context
- KHÔNG bịa đặt hoặc thêm thông số kỹ thuật không có trong dữ liệu
- KHÔNG đề cập sản phẩm không có trong thông tin được cung cấp
- Trả lời bằng tiếng Việt, ngắn gọn, rõ ràng, dễ hiểu
- Luôn đề cập TÊN SẢN PHẨM và GIÁ CỤ THỂ khi giới thiệu
- Giải thích TẠI SAO gợi ý sản phẩm đó (ưu điểm phù hợp với nhu cầu)
- Sử dụng emoji phù hợp để làm rõ ý (⭐ 💻 🎮 💰 ✅ ❌)
- Nếu không có thông tin đủ để trả lời, hãy thừa nhận và hỏi thêm
- Khi giới thiệu sản phẩm, đề cập rằng khách hàng có thể click vào card sản phẩm bên dưới để xem chi tiết

CÁCH TRẢ LỜI:
- Với câu hỏi chung: Gợi ý 2-3 sản phẩm phù hợp nhất
- Với câu hỏi cụ thể: Trả lời trực tiếp dựa vào thông tin
- Với so sánh: Đưa ra bảng so sánh hoặc phân tích từng điểm
- Cuối mỗi câu trả lời: Hỏi thêm để hiểu rõ nhu cầu
- Luôn nhắc khách "Click vào sản phẩm bên dưới để xem đầy đủ thông tin và đặt hàng"`;
  }

  /**
   * Generate response từ LLM với context từ RAG
   */
  async generateResponse(userMessage, context, conversationHistory = []) {
    const startTime = Date.now();

    try {
      // Build messages
      const messages = this.buildMessages(userMessage, context, conversationHistory);

      console.log('🤖 Generating response with LLM...');

      // Call HuggingFace Inference API
      let fullResponse = '';

      const stream = await this.hf.chatCompletionStream({
        model: this.model,
        messages: messages,
        max_tokens: 800,
        temperature: 0.7,
        top_p: 0.9
      });

      // Collect streamed response
      for await (const chunk of stream) {
        if (chunk.choices && chunk.choices.length > 0) {
          const delta = chunk.choices[0].delta;
          if (delta.content) {
            fullResponse += delta.content;
          }
        }
      }

      const generationTime = Date.now() - startTime;
      console.log(`✅ LLM response generated in ${generationTime}ms`);

      // Estimate tokens (rough approximation)
      const estimatedTokens = Math.ceil(fullResponse.length / 4);

      return {
        content: fullResponse.trim(),
        model: this.model,
        generation_time_ms: generationTime,
        tokens_used: estimatedTokens
      };
    } catch (error) {
      console.error('❌ LLM generation error:', error);

      // Fallback response nếu LLM fail
      return {
        content: this.generateFallbackResponse(context),
        model: 'fallback',
        generation_time_ms: Date.now() - startTime,
        tokens_used: 0,
        error: error.message
      };
    }
  }

  /**
   * Build messages array cho LLM
   */
  buildMessages(userMessage, context, conversationHistory) {
    const messages = [
      {
        role: 'system',
        content: this.systemPrompt
      }
    ];

    // Add conversation history (last 5 messages)
    const recentHistory = conversationHistory.slice(-5);
    recentHistory.forEach((msg) => {
      messages.push({
        role: msg.role,
        content: msg.content
      });
    });

    // Add current user message với context
    messages.push({
      role: 'user',
      content: `${context}\n\n---\n\nCâu hỏi của khách hàng: ${userMessage}`
    });

    return messages;
  }

  /**
   * Generate fallback response khi LLM fail
   */
  generateFallbackResponse(context) {
    if (context.includes('Không tìm thấy')) {
      return 'Xin lỗi, tôi không tìm thấy sản phẩm phù hợp với yêu cầu của bạn. Bạn có thể mô tả chi tiết hơn về nhu cầu sử dụng và ngân sách của mình không?';
    }

    // Extract product info from context (simple parsing)
    const productMatches = context.match(/\[Sản phẩm \d+\] (.+)\nGiá: (.+)/g);

    if (productMatches && productMatches.length > 0) {
      let response = 'Dựa vào yêu cầu của bạn, tôi tìm thấy các sản phẩm sau:\n\n';

      productMatches.forEach((match, idx) => {
        const nameMatch = match.match(/\[Sản phẩm \d+\] (.+)/);
        const priceMatch = match.match(/Giá: (.+)/);

        if (nameMatch && priceMatch) {
          response += `${idx + 1}. ${nameMatch[1]} - ${priceMatch[1]}\n`;
        }
      });

      response += '\nBạn muốn biết thêm thông tin chi tiết về sản phẩm nào?';
      return response;
    }

    return 'Xin lỗi, tôi đang gặp sự cố kỹ thuật. Vui lòng thử lại sau ít phút hoặc liên hệ bộ phận hỗ trợ.';
  }

  /**
   * Validate API key
   */
  validateApiKey() {
    if (!process.env.HUGGINGFACE_API_KEY) {
      throw new Error('HUGGINGFACE_API_KEY is not configured in .env file');
    }
    return true;
  }
}

export default new LlmService();

File: ai-service/app/services/llm_service.py
import os
import time
import logging
from typing import List, Dict, Any, Optional
import requests

from app.core.config import settings

logger = logging.getLogger("ai-service.llm_service")

class LLMService:
    def __init__(self):
        self.opencode_api_key = settings.OPENCODE_API_KEY
        self.opencode_base_url = settings.OPENCODE_BASE_URL.rstrip("/")
        self.opencode_model = settings.OPENCODE_MODEL or "deepseek-v4-flash-free"
        self.openai_api_key = settings.OPENAI_API_KEY
        self.hf_api_key = settings.HUGGINGFACE_API_KEY
        self.hf_model = "Qwen/Qwen2.5-7B-Instruct"

        self.system_prompt = """Bạn là một chuyên gia tư vấn laptop công nghệ nhiệt tình, thông minh và gần gũi (như một người bạn am hiểu phần cứng đang tư vấn tận tâm cho người mua).

PHONG CÁCH VÀ NGUYÊN TẮC TƯ VẤN:
1. Giao tiếp tự nhiên, thân thiện, xưng hô "mình" và "bạn". Tuyệt đối KHÔNG trả lời dập khuôn như robot hay biểu mẫu cứng nhắc.
2. NẮM BẮT TRỌNG TÂM & KẾ THỪA LỊCH SỬ HỘI THOẠI:
   - Bạn PHẢI ghi nhớ các điều kiện người dùng đã đề cập ở các câu trước (ví dụ: ngân sách 20-30 triệu, thương hiệu, hoặc mục đích sử dụng).
   - Khi người dùng hỏi thêm hoặc tập trung vào một chi tiết cụ thể (ví dụ: màn hình độ phân giải cao, chip CPU khỏe, card đồ họa rời, thời lượng pin, trọng lượng máy): Bạn hãy TRỰC DIỆN trả lời và phân tích sâu đúng chi tiết đó trước!
   - Ví dụ: Khách hỏi "có độ phân giải cao không" -> Hãy trả lời ngay: "Trong tầm giá bạn đang chọn, mình thấy các mẫu này sở hữu màn hình rất ấn tượng: máy A trang bị màn 2K/OLED sắc nét, máy B có tần số quét 144Hz mượt mà..."
   - Nếu khách hỏi "cpu cao" -> Phân tích điểm mạnh CPU (i7/Ryzen 7 đa nhân xử lý mượt mà tác vụ nặng).
3. PHẠM VI SẢN PHẨM & THƯƠNG HIỆU CỬA HÀNG:
   - Cửa hàng hiện phân phối chính hãng 4 thương hiệu laptop: Dell, Lenovo, Acer, MSI.
   - Cửa hàng KHÔNG KINH DOANH Apple (MacBook), Asus, HP, Samsung, v.v.
   - Khi khách hỏi dòng máy không bán (như Apple/MacBook): Hãy lịch sự thông báo cửa hàng chưa kinh doanh dòng này, và khéo léo giới thiệu các lựa chọn tương đương có sẵn trong kho (ví dụ: Lenovo ThinkPad/Yoga, Dell Inspiron/XPS, Acer Swift cho phân khúc mỏng nhẹ cao cấp). Tuyệt đối không tự bịa thông số máy Apple.
4. CÂU HỎI NGOÀI PHẠM VI (OUT OF SCOPE):
   - Nếu khách hỏi thời tiết, nấu ăn, làm thơ, chính trị...: Lịch sự từ chối ngắn gọn và kéo câu chuyện quay lại nhu cầu tư vấn laptop.
5. CHÍNH SÁCH CỬA HÀNG (BẢO HÀNH, ĐỔI TRẢ, GIAO HÀNG, TRẢ GÓP):
   - Trả lời đầy đủ, chuẩn xác theo thông tin hệ thống cung cấp trong context.
6. ĐÁNH GIÁ SẢN PHẨM THỰC TẾ:
   - Không liệt kê thông số kỹ thuật khô khan như bảng mã. Hãy giải thích lợi ích thực tế của linh kiện đối với trải nghiệm của khách.
   - Nhắc đến tên máy và khoảng giá một cách tự nhiên.
7. Emoji sinh động, vừa phải (✨, 💻, 🚀, 🎯, 🔥).
8. Nếu có sản phẩm gợi ý, cuối câu trả lời nhắc nhẹ bạn ấy xem nhanh các thẻ sản phẩm kèm ảnh và link chi tiết ở ngay bên dưới."""

    def generate_response(
        self,
        user_message: str,
        context: str,
        conversation_history: List[Dict[str, Any]] = None
    ) -> Dict[str, Any]:
        start_time = time.time()
        conversation_history = conversation_history or []

        # 1. Try OpenCode Zen / DeepSeek API (OpenAI-compatible) with low timeout (2.5s)
        if self.opencode_api_key:
            candidate_models = ["ling-3.0-flash-fin-free"]

            messages = [{"role": "system", "content": self.system_prompt}]
            for msg in conversation_history[-4:]:
                messages.append({"role": msg.get("role", "user"), "content": msg.get("content", "")})
            messages.append({"role": "user", "content": f"Danh sách sản phẩm trong kho:\n{context}\n\n---\nCâu hỏi của khách hàng: {user_message}"})

            for model_name in candidate_models:
                try:
                    res = requests.post(
                        f"{self.opencode_base_url}/chat/completions",
                        headers={
                            "Authorization": f"Bearer {self.opencode_api_key}",
                            "Content-Type": "application/json"
                        },
                        json={
                            "model": model_name,
                            "messages": messages,
                            "temperature": 0.7,
                            "max_tokens": 400
                        },
                        timeout=2.5
                    )
                    if res.status_code == 200:
                        data = res.json()
                        choices = data.get("choices") or []
                        if choices:
                            msg_obj = choices[0].get("message", {})
                            content = msg_obj.get("content") or ""
                            if content and content.strip():
                                elapsed = (time.time() - start_time) * 1000
                                logger.info(f"Generated response using OpenCode model: {model_name}")
                                return {
                                    "content": content.strip(),
                                    "model": f"opencode/{model_name}",
                                    "generation_time_ms": elapsed,
                                    "tokens_used": data.get("usage", {}).get("total_tokens", 0)
                                }
                except Exception as e:
                    logger.warning(f"OpenCode Zen ({model_name}) request failed or timed out: {e}")

        # 2. Try OpenAI if API key configured
        if self.openai_api_key:
            try:
                import openai
                client = openai.OpenAI(api_key=self.openai_api_key)
                messages = [{"role": "system", "content": self.system_prompt}]
                for msg in conversation_history[-4:]:
                    messages.append({"role": msg.get("role", "user"), "content": msg.get("content", "")})
                messages.append({"role": "user", "content": f"Danh sách sản phẩm trong kho:\n{context}\n\n---\nCâu hỏi của khách hàng: {user_message}"})

                response = client.chat.completions.create(
                    model="gpt-3.5-turbo",
                    messages=messages,
                    temperature=0.7,
                    max_tokens=500,
                    timeout=2.5
                )
                content = response.choices[0].message.content
                elapsed = (time.time() - start_time) * 1000
                return {
                    "content": content,
                    "model": "gpt-3.5-turbo",
                    "generation_time_ms": elapsed,
                    "tokens_used": response.usage.total_tokens if response.usage else 0
                }
            except Exception as e:
                logger.warning(f"OpenAI completion failed: {e}")

        # 3. Fast Intelligent Dynamic Fallback Generator (Zero latency < 5ms)
        elapsed = (time.time() - start_time) * 1000
        fallback_text = self._build_dynamic_response(user_message, context)
        return {
            "content": fallback_text,
            "model": "smart_rule_engine",
            "generation_time_ms": elapsed,
            "tokens_used": 0
        }

    def _build_dynamic_response(self, query: str, context: str) -> str:
        # 1. Direct System Notifications from Context
        if "THÔNG BÁO HỆ THỐNG (OUT OF SCOPE)" in context:
            return (
                "Chào bạn! 👋 Hiện tại mình là trợ lý ảo chuyên hỗ trợ tư vấn chọn mua laptop và thiết bị phần cứng công nghệ. "
                "Chủ đề này hơi nằm ngoài chuyên môn của mình rồi 😄. Bạn có đang cần tư vấn hoặc tìm mẫu laptop phục vụ học tập, làm việc hay chơi game không, để mình hỗ trợ nhé! ✨"
            )

        if "THÔNG BÁO HỆ THỐNG (STORE POLICY)" in context:
            return (
                "Dạ chào bạn! Cửa hàng xin gửi tới bạn thông tin chính sách bảo hành và dịch vụ mới nhất ạ:\n\n"
                "🛡️ **Chính sách bảo hành:** 100% sản phẩm là hàng chính hãng, bảo hành từ 12 - 24 tháng theo đúng chuẩn NSX.\n"
                "🔄 **Chính sách đổi trả:** 1 đổi 1 trong 30 ngày đầu tiên nếu máy có lỗi phần cứng do nhà sản xuất.\n"
                "🚚 **Giao hàng:** Miễn phí vận chuyển toàn quốc, hỗ trợ kiểm tra máy trước khi thanh toán (COD).\n"
                "💳 **Hỗ trợ trả góp:** Áp dụng trả góp 0% lãi suất qua thẻ tín dụng và các đối tác tài chính uy tín.\n\n"
                "Nếu bạn cần thêm thông tin chi tiết về dòng máy nào, cứ nhắn mình hỗ trợ nhé! 💡"
            )

        if "THÔNG BÁO HỆ THỐNG (THƯƠNG HIỆU KHÔNG KINH DOANH)" in context:
            return (
                "Chào bạn! 👋 Hiện tại cửa hàng chúng mình **không kinh doanh dòng Apple (MacBook)** mà tập trung phân phối chính hãng 4 thương hiệu hàng đầu: **Dell, Lenovo, Acer, MSI**.\n\n"
                "Nếu bạn yêu thích phong cách thiết kế sang trọng, mỏng nhẹ, pin tốt và màn hình đẹp như MacBook, cửa hàng có các dòng tương đương cực kỳ xuất sắc như:\n"
                "- **Lenovo ThinkPad / Yoga:** Bàn phím đỉnh cao, bền bỉ, màn hình OLED 2.8K siêu đẹp.\n"
                "- **Dell XPS / Inspiron cao cấp:** Vỏ nhôm nguyên khối tinh tế, hiệu năng mạnh mẽ.\n"
                "- **Acer Swift AI:** Siêu nhẹ ~1.2kg, chuẩn Intel Evo/Core Ultra mượt mà.\n\n"
                "Bạn có muốn tham khảo các mẫu máy này trong tầm giá mong muốn không ạ? ✨"
            )

        if not context or "Không tìm thấy" in context:
            return (
                "Xin chào! Hiện tại mình chưa tìm thấy mẫu laptop nào khớp hoàn toàn với tất cả tiêu chí của bạn trong kho hàng. "
                "Bạn có thể điều chỉnh lại một chút khoảng giá hoặc cho mình biết thêm nhu cầu cụ thể (ví dụ: làm việc văn phòng, lập trình hay chơi game) để mình hỗ trợ nhé! 💡"
            )

        q_low = query.lower()
        intent_greeting = "Dựa trên nhu cầu của bạn, mình xin đề xuất các mẫu laptop phù hợp nhất đang có sẵn tại cửa hàng:"
        if any(w in q_low for w in ["độ phân giải", "màn hình", "2k", "4k", "oled", "120hz", "144hz"]):
            intent_greeting = "Về tiêu chí **màn hình và độ phân giải cao**, mình đã lọc ngay các dòng máy có màn hình sắc nét vượt trội (OLED, 2K+, tần số quét cao) trong tầm giá cho bạn:"
        elif any(w in q_low for w in ["cpu", "chip", "i7", "i9", "ryzen 7", "ryzen 9", "cpu cao", "chip mạnh"]):
            intent_greeting = "Về yêu cầu **CPU hiệu năng cao**, mình đã chọn các mẫu sở hữu vi xử lý đa nhân mạnh mẽ hàng đầu (Core i7/i9 hoặc Ryzen 7/9) tối ưu cho các tác vụ nặng:"
        elif any(w in q_low for w in ["lập trình", "code", "dev", "it", "phần mềm"]):
            intent_greeting = "Dành cho nhu cầu **Lập trình & Phát triển phần mềm**, mình đã chọn lọc các mẫu laptop có CPU khỏe, dung lượng RAM và ổ cứng tối ưu trong tầm giá:"
        elif any(w in q_low for w in ["game", "gaming", "đồ họa", "render", "3d", "card rời", "gpu"]):
            intent_greeting = "Dành cho nhu cầu **Gaming & Đồ họa**, dưới đây là những mẫu laptop trang bị Card đồ họa rời (GPU) và tản nhiệt tối ưu nhất:"
        elif any(w in q_low for w in ["văn phòng", "học tập", "sinh viên", "mỏng nhẹ"]):
            intent_greeting = "Dành cho nhu cầu **Học tập & Làm việc văn phòng**, đây là những lựa chọn mỏng nhẹ, pin tốt và giá cả cực kỳ hợp lý:"

        # Format products naturally
        intro = f"Chào bạn! 👋 {intent_greeting}\n\n"
        body = f"{context}\n\n" if context else ""
        footer = "💡 Bạn có thể bấm trực tiếp vào thẻ sản phẩm bên dưới để xem hình ảnh chi tiết và đặt hàng nhé! ✨"
        return f"{intro}{body}{footer}"

llm_service = LLMService()
